File: nautile/shaders/fx/Ocean.fs
/** 
The MIT License (MIT)

Copyright (c) 2016 Frederic Condolo

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
**/

#define	RAYMARCH_ITERATION_HOP		s += f*0.003;
#define	RAYMARCH_STEP_SCALE			0.99
#define RAYMARCH_EPSILON 			0.001
#define	RAYMARCH_MAX_ITER 			32
#define	RAYMARCH_MAX_DIST			10.0
#define RAYMARCH_OVERLOAD_COMPUTE_COLOR
#define	RAYMARCH_RAY_START_MIN_DIST 0.25

//#define USE_MASKING
#define SCENE myscene(_p, _d)
#define MIN_CAM_DIST 0.0 // 1.0
#define MAX_ITER 64
#define COLLISION_EPSILON 0.1
#define START_DIST 3.0
#define COLORIZE computeColor
#define START_FX_FRAME startFrame();


#define CAMERA_LENS_DISTORTION drops = texture2D( uSampler5, rmdata.frag + vec2(1.0, iGlobalTime*0.011)); vec4 waternormal = clamp(cos(0.55*iGlobalTime+(rmdata.frag.x+rmdata.frag.y)*10.0), 0.0, 0.5)*pow(1.0-rmdata.frag.y,2.0)*(drops); rmdata.dir = normalize(rmdata.dir + 1.0/(1.0*length(waterDrops.yz-(rmdata.frag))+1.0)*waterDrops.x*waternormal.xyz);

//parameters
#define FOG_END 20.0
uniform vec2 noise1Spd;
uniform vec2 noise2Spd;
uniform vec3 ScrollSpd;
uniform float noise1Base;
uniform float noise2Scale;
uniform float noise2Base;
uniform float noise1Scale;
uniform float waveHeightSpd;
uniform float waveHeightStrength;
uniform float minWaveHeight;
uniform float noise1W;
uniform float noise2W;
uniform vec4 underLightPos;
uniform vec3 waterDrops;

#define COLLISION_TEST
#define SKIP_FARTEST
#define FOAM_SPD 0.25

// local variables
vec4 drops;
float altitude;
vec3 colorAccu;
float colorAmount;
float iterCount;
vec3 clouds;



vec3 computeColor(Hit hit) {
    rmdata.globalCol = colorAccu/colorAmount + max(max(drops.x,drops.y), drops.z) * 0.051;
	return rmdata.globalCol*1.25;
}


// Noise functions...
float Hash( float n )
{
	return  texture2D( uSampler1, vec2(n*.27, n*.26), -100.0 ).x;
}

//--------------------------------------------------------------------------
vec2 Hash2(float n)
{
	return texture2D( uSampler1, vec2(n*.077331, n*.066927), -100.0 ).xz;
}


float noise( in vec2 x )
{
    vec2 p = floor(x);
    vec2 f = fract(x);
    f = f*f*(3.0-2.0*f);
    float n = p.x + p.y*57.0;
    float res = mix(mix( Hash(n+  0.0), Hash(n+  1.0),f.x),
                    mix( Hash(n+ 57.0), Hash(n+ 58.0),f.x),f.y);
    return res;
}
float FractalNoise(in vec2 xy)
{
	float w = noise1W;
	float f = 0.0;
	float norm = 0.0;
	float mult = 0.4*(1.0+0.3*abs(sin(rmdata.p.z)))+abs(sin(rmdata.p.x*0.01));
	for (int i = 0; i < 4; i++)
	{
		f += noise(xy) * w;
		norm += w;
		w *= 0.5;
		xy *= mult;
	}
	return f / norm;
}

float FractalNoise2(in vec2 xy, in float n1)
{
	float w = noise2W * abs(n1);
	float f = 0.0;
    //float mult = 0.4*(1.0+0.3*abs(cos(rmdata.p.z*1.2)));
	float mult = noise2Base+noise2Scale*(abs(cos(rmdata.p.z*1.2))+abs(sin(0.04*iGlobalTime*noise2Spd.x+xy.x*0.05))*abs(cos(0.04*iGlobalTime*noise2Spd.y+xy.y*0.05)));
	float norm = 0.0;
	for (int i = 0; i < 4; i++)
	{
		f += noise(xy) * w;
		norm += w;
		w *= 0.5;
		xy *= mult;
	}
	return f / norm;
}

float terrainH( vec3 p )
{
    float wavescale = 1.5;
	vec2 pos = wavescale*vec2(p.x*0.5, p.z*0.75 + iGlobalTime*0.025*FOAM_SPD);
	float n1 = FractalNoise(pos);
	float n2 = FractalNoise2(pos, n1);
	n1 = clamp(max(0.0,n1-0.75)*6.0, 0.0, 0.99999);
	n2 = clamp(max(0.0,n2-0.75)*6.0, 0.0, 0.99999);
	altitude = (n2*0.5+n1)*0.5;
    float sum = 10.0*altitude ;//* (3.0+cos(-iGlobalTime*waveHeightSpd+p.z+iterCount*waveHeightStrength));
    return sum;
}


//Object
float myscene(in vec3 p, in float f)
{ 		
    float maxDvalue = 1.0;
    float cylRad = 5.0;
    float maxsd = 35.0;
    vec3 relcam = p-rmdata.camera;
	float sd = sdCylinder(vec3(relcam.z+cylRad*0.5*(1.0+sin(-0.25*iGlobalTime+relcam.z*0.2)), relcam.y, relcam.x+cylRad*0.5*(1.0+cos(3.14-0.25*iGlobalTime+relcam.z*0.3))), vec3(0.051,1.0,0.051));
    sd = min(sd*2.8,maxsd);
	vec3 p2 = vec3(p.x, p.y-sd, p.z);
//	vec3 p2 = p;
	float d = 1.1*altitude+p2.y-terrainH(p2);
 //   if (d<10.0) {
		colorAmount += (0.075 * sd+1.0);
        vec4 foam = texture2D( uSampler0, vec2(p.x, p.z+FOAM_SPD*iGlobalTime)*vec2(0.15+0.05*cos(p.z),0.4));
        foam.rgb = pow(vec3(1.0+foam.rgb), vec3(4.0)) * 0.15;
        altitude *= 1.0+rmdata.dist*0.08*(1.0-foam.b)/max(1.0,altitude+(rmdata.p.z-rmdata.camera.z)*0.5);
        altitude *= 2.75;
		//colorAccu += dFactor*(foam.rgb*30.0/rmdata.dist)*0.5*
     //   vec3 baseCol = texture2D( uSampler2, vec2(clamp(0.0, 1.0, altitude / max(f/FOG_END,4.0)), 1.0) ).xyz;
        altitude = max(altitude,0.3);
        vec3 baseCol = vec3(clamp(0.4, 1.0, altitude / max(f/FOG_END,4.0)));
        colorAccu += 7.0*vec3(0.7,0.8,1.0)*(baseCol*foam.rgb);
/*
        vec3 lp = underLightPos.xyz;
//        float dist = length(lp.xyz-(rmdata.p.xyz-rmdata.camera.xyz));
        float dist = length(lp.xz-(rmdata.p.xz-rmdata.camera.xz));
        dist = 1.0-pow(clamp(dist,0.0,1.0), 2.0);
        colorAccu += 0.5*dist*vec3(0.6,0.6,1.0);
        */
//	}
	return d*0.1;//minWaveHeight+d*0.25;
}

void startFrame() {
		rmdata.globalCol = vec3(0.0, 0.0, 0.0);
        colorAccu = vec3(0.0,0.0,0.0);

        colorAmount = 0.0;
        clouds = vec3(0.0,0.0,0.0);
        iterCount = 0.0;
}

File: nautile/js/fx/Ocean.js
/** 
The MIT License (MIT)

Copyright (c) 2016 Frederic Condolo

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
**/


/**
 * @constructor
 */
function  Ocean(params) {
    var t = this;
    t.params = params;
    t.hlp = new RaymarchFXHelper();
    t.hlp.construct({
        name: "Ocean",
        parent: t,
        prevFramesCount: 2,
        fragmentShadersPasses: [{
            id: "final",
            shader: "fx/Ocean.fs",
            inputs: null,
            outputTex: { // for fields description, check createRenderTexture in engine/Texture.js
                width: window.innerWidth,
                height: window.innerHeight
            }
        },
            {
                id: "horizBlur",
                disableRaymarch: true,
                shader: "engine/HorizBlurVignette.fs",
                inputs: [{ id: "final", sampler: 1 }],
                outputTex: { // for fields description, check createRenderTexture in engine/Texture.js
                    width: window.innerWidth,
                    height: window.innerHeight
                }
            },
            {
                id: "vertBlur",
                disableRaymarch: true,
                shader: "engine/VertBlurVignette.fs",
                inputs: [ { id: "horizBlur", sampler: 1 }],
                outputTex: { // for fields description, check createRenderTexture in engine/Texture.js
                    width: window.innerWidth,
                    height: window.innerHeight
                }
            },
            {
                id: "logo",
                disableRaymarch: true,
                shader: "fx/OceanLogo.fs",
                inputs: [{ id: "vertBlur", sampler: 1 }],
                outputTex: null
            }
        ]
    });
    t.interpParams = [];
    t.textures = [];
    t.textureParams = [];
    t.underLightPos = [];
    t.waterDrops = [];
    t.allParams = {};
    t.remaining = 100.0;
    t.DOF = 0.05;

    // Camera
    t.prevCam = [0.0, 0.0, 0.0];
    t.neon = 0.0;
    t.neon2 = 0.0;
    t.logoAlpha = 0.0;
    t.time = 0.0;
    t.ending = false;
}

Ocean.prototype = {
	
    loadTextures: function (_reload) {
        var t = this;
        engine.pushContext(webgl_2d_raymarch, "Ocean preloadResources");
        for (var i = 0; i < t.textureParams.length; i++)
            t.textures[i] = resman.prefetchTexture(t.textureParams[i].file);
        engine.popContext();
    },

    preloadResources: function () {
        var t = this;
        t.hlp.preloadResources();
        t.loadTextures(false);
    },

    exit: function () {
        var t = this;
        t.hlp.exit();
        for (var i = 0; i < t.textures.length; i++)
            t.textures[i].release();
        engine.text2d.closeAllIn(2.5);
      //  engine.text2d.clearAll();
    },

    enter: function () {
        var t = this;
        t.hlp.enter();
        t.exitRectDone = false;
   //     t.text2b = { x: 0.01, y: 0.17, string: "Nautile", fillStyle: "#FFFFFF", alpha: 0.0, fadeIn: { wait: 6.0, duration: 0.01 }, fadeOut: { wait: 3.0, duration: 4.0, xspd: -0.05 }, max:0.5 };
        //t.text2 = { x: 0.01, y: 0.05, string: "Present", fillStyle: "#FFFFFF", alpha: 0.0, fadeIn: { wait: 3.0, duration: 3.0, yspd: 0.05 }, fadeOut: { wait: 3.0, duration: 4.0, xspd: 0.025 }, max: 0.8 };
        t.text1 = { x: 0.01, y: 0.05, string: "Deadliners", fillStyle: "#FFFFFF", alpha: 0.0, fadeIn: { wait: 8.0, duration: 3.0, yspd: 0.05 }, fadeOut: { wait: 3.0, duration: 5.0, xspd: 0.025 }, max: 0.8 };
        engine.addCanvasClient(engine.frontCanvas, "Ocean.enter");
        engine.text2d.addEntry(t.text1);
        engine.text2d.addRectangle({ x: 0.0, y: 0.0, w: 1.0, h: 1.0, a: 0.0, done: false, fillStyle: "#000000", alpha: 1.0, fadeOut: { wait: 0.0, duration: 2.0 } });

        engine.pushContext(webgl_2d_raymarch, "Ocean enter");
            var prg = t.hlp.getFragmentShaderPass("final").program;
            prg.underLightPos = engine.gl.getUniformLocation(prg, "underLightPos");
            prg.waterDrops = engine.gl.getUniformLocation(prg, "waterDrops");

            prg = t.hlp.getFragmentShaderPass("logo").program;
            prg.neon = engine.gl.getUniformLocation(prg, "neon");
            prg.neon2 = engine.gl.getUniformLocation(prg, "neon2");
            prg.logoAlpha = engine.gl.getUniformLocation(prg, "logoAlpha");

            prg = t.hlp.getFragmentShaderPass("horizBlur").program;
            prg.sensitivity = engine.gl.getUniformLocation(prg, "sensitivity");

            prg = t.hlp.getFragmentShaderPass("vertBlur").program;
            prg.sensitivity = engine.gl.getUniformLocation(prg, "sensitivity");


        engine.popContext();
        t.remaining = 100.0;
        t.time = 0.0;
    },

    manualCamera: function () {
        var t = this;
        var suite = t.parent;
        suite.campos[0] = 3.0 * Math.cos(t.hlp.elapsed * 0.3);
        var posY = Math.max(23.0,40.0 - t.hlp.elapsed);
        suite.campos[1] = posY;
        if (t.remaining < 10.0)
            suite.campos[1] += 1.0*(10.0 - t.remaining);
        suite.campos[2] = t.prevCam[2] + 0.4 * engine.deltaTime;
        suite.camYaw = 0.0;//-1.1 * Math.cos(t.hlp.elapsed * 0.1);
        suite.camRoll = 0.0;////-0.5*Math.cos(t.hlp.elapsed*0.1);
        suite.camPitch = 1.4;
        t.prevCam[0] = suite.campos[0];
        t.prevCam[1] = suite.campos[1];
        t.prevCam[2] = suite.campos[2];
    },

    executeRenderPass: function (_passName) {
        var t = this;
        if (!t.hlp.enter_tick(_passName))
            return;

        var gl = engine.gl;
        var pass = t.hlp.getFragmentShaderPass(_passName);
        var prg = pass.program;

        engine.setTexture(t.textures[0], 1); // noise
        engine.setTexture(t.textures[1], 0); // caustics
        engine.setTexture(t.textures[2], 5, true); // water drops

        if (prg.underLightPos) engine.gl.uniform4fv(prg.underLightPos, t.underLightPos);
        engine.gl.uniform3fv(prg.waterDrops, t.waterDrops);

        t.hlp.exit_tick(_passName);
    },

    executePostProcessPass: function (_passName) {
        var t = this;
        if (!t.hlp.enter_tick(_passName))
            return;

        var prg = t.hlp.getFragmentShaderPass(_passName).program;

        if (prg.lightPos)
            engine.gl.uniform2fv(prg.lightPos, [t.underLightPos[0], t.underLightPos[1]]);
        if (prg.decay)
            engine.gl.uniform1f(prg.decay, 0.96815);

        if (prg.sensitivity)
            engine.gl.uniform1f(prg.sensitivity, t.DOF);

        if (prg.neon) {
            t.neon += ((Math.random() * 0.5) - t.neon) * 0.25;
            if (t.remaining < 7.0)
                t.neon2 += ((Math.random()) - t.neon2) * 0.5;
            if (t.time > 24.0)
                t.logoAlpha = Math.min(1.0, (t.time - 22.0) * 0.1);
            if (t.remaining < 5.0)
                t.logoAlpha = Math.max(0.0, (t.remaining + 5.0) * 0.1);
        //    t.logoAlpha = 1.0;
            if (t.remaining < 2.0 && beatHandler.beatOccuredThisFrame) {
                t.ending = true;
            }
            if (t.ending) {
                t.neon = 4.0;
                t.neon2 = 4.0;
             //   t.logoAlpha = 0.0
            }
            engine.gl.uniform1f(prg.logoAlpha, t.logoAlpha);
            engine.gl.uniform1f(prg.neon, t.neon);
            engine.gl.uniform1f(prg.neon2, t.neon2);
        }

        engine.setTexture(t.textures[3], 2, true); // logo
        engine.setTexture(t.textures[4], 4, true); // flare

        t.hlp.exit_tick(_passName);
    },

    tick: function (time, _gl, remaining) {
        var t = this;
        t.DOF = 0.15*Math.abs(Math.sin(engine.time)) + 0.01;
        t.remaining = remaining;
        var deltaRemaining = 1.0;
        if (remaining < 10.0) {
            deltaRemaining = Math.max(1.0, 10.0 - remaining);
            if (remaining < 2.0 && !t.exitRectDone && beatHandler.beatOccuredThisFrame) {
                t.exitRectDone = true;
                engine.text2d.addRectangle({ x: 0.0, y: 0.0, w: 1.0, h: 1.0, a: 0.0, done: false, fillStyle: "#FFFFFF", alpha: 0.0, fadeIn: { wait: 0.0, duration: 0.5 }, fadeOut: { wait: 0.0, duration: 1.0 } });
            }
        }


        t.underLightPos[0] = Math.cos(time * 0.9) * (2.5/deltaRemaining);
        t.underLightPos[1] = t.parent.campos[1]-2.0;//0.0;//Math.sin(time * 0.9) * 0.2;
        t.underLightPos[2] = remaining + Math.abs(Math.sin(time * 0.4) * 0.1);
        t.underLightPos[3] = Math.max(15.0,30.0-deltaRemaining);

        t.waterDrops[0] = Math.min(1.0,Math.max(0.0, 2.0*Math.sin(time*1.5)));
        t.waterDrops[1] = 2.0*Math.cos(time * 0.9) + 1.0;
        t.waterDrops[2] = 2.0 * Math.sin(time * 0.9) + 1.0;

        t.executeRenderPass("final");
        t.executePostProcessPass("horizBlur");
        t.executePostProcessPass("vertBlur");
        t.executePostProcessPass("logo");

        t.time += engine.deltaTime;


        return true;
    },


    createFX : function() {
        var t = this;
        t.hlp.createFX();
    }
}
